### DISCUSSION TITLE
Braindump of Ideas

### DISCUSSION THREAD

**Discussion by PetalCat** [PSDTools/app#13] [Ideas]
> like https://github.com/gatsbyjs/gatsby/issues/1
 
The place to discuss future features and plans.[^1] If you have an idea randomly while developing, just throw it in here. This is not a for sure to-do list—just a bunch of ideas for after we have the Pirate Wallet framework built... then, if you're bored, you can attempt to implement something from this list.

- Anyone is free to comment on this to add ideas.
- Folks who can edit will add it.
- Just don't remove ideas already here.
- Even if it's a bad idea it might still be a good inspiration.

## Applets:
- [ ] #191
- [ ] #4
- [x] #2
- [ ] #3
- [ ] #190
- [ ] Applet: Teacher Data
  - **Description**
    - An applet on the teacher flavor of the app that allows teachers to pull data from multiple different platforms and use it all in one space.[^2]
  - **Details**
    - platforms including
      - PowerSchool,
      - Canvas,
      - Zipgrade
      - (add more if you can think of any)
- [ ] Applet: News
  - **Description**
    - A news applet that shows current events at the school.
  - **Details**
    - Might be interesting to make teachers able to blog?[^3] (make news articles)
    - Make a feed of all of the news articles
    - It would be interesting to have different classes publish writing and media to it.
    - Clubs could also post
    - Be able to select what types of content show up
- [ ] Applet: Sports
    - **Description**
      - Show sports scores after games
- [ ] Applet: Clubs
    - **Description**
      - Some sort of applet with club info (the only change from the current website would be convenience)

## Features:

- [ ] Feature: Notifications
  - **Details**
    - Get notified of:
      - closures
      - types of events the user enables

[^1]: I (@PetalCat) think we should call the different apps in the Wallet Applets unless someone has better terminology
[^2]: I was talking to Mrs. Bulger today about how useful it would be to connect all the data like that and maybe even set up a system where teachers could make "programs" sort of like Google Sheets to automatically calculate different things using all of that data that all in one place
[^3]: What if we built a free/libre open source software framework for the sake of making a nice-to-use API we could hook up with a custom Content Management System for easy blogging for the [psdr3 website](https://psdr3.org/), [PattonvilleToday](https://pattonvilletoday.com/), and any other school websites?[^4]
[^4]: The [Pirate Players](https://www.pirateplayers.org/) is the other one I know of, but I'm sure there's more.

**Comment by lishaduck** [comment]
Flutter packages to look into:

> - <https://codelabs.developers.google.com/codelabs/flutter-codelab-first#0>
> - <https://pub.dev/packages/go_router>
> - <https://pub.dev/packages/flutter_platform_widgets>

_Originally posted by @lishaduck in https://github.com/PSDTools/app/issues/14#issuecomment-1518426486_

**Reply by lishaduck** [reply]
## UPDATE:
- Our router is [`go_router`](https://pub.dev/packages/go_router) because it's Flutter's official routing package.
  - I'm going to set up typesafe routes with [`go_router_builder`](https://pub.dev/packages/go_router_builder) soon if you don't mind having to use Code Generation.[^1]
  - If you need a link, keep in mind that you should use [`url_launcher`](https://pub.dev/packages/url_launcher) because it creates an actual `<a>` HTML element in web mode.
- We're currently using [`provider`](https://pub.dev/packages/provider), but I think we'll switch to [`riverpod`](https://docs-v2.riverpod.dev/).

_**NOTE**: Only `go_router` is currently installed, so if you want to work on it once I push my changes, you'll need to use `flutter pub add`._

[^1]: which means we can use Riverpod easier also because v2 uses optional code generation

**Reply by lishaduck** [reply]
This looks useful: <https://github.com/ookami-kb/storybook_flutter>.

**Reply by lishaduck** [reply]
https://www.widgetbook.io/

**Reply by lishaduck** [reply]
<https://dcm.dev/>

Oh, and to put it on paper, we use auto_router, not go_router.

**Reply by lishaduck** [reply]
If we want to break stuff into multiple packages, it looks like [Melos](https://melos.invertase.dev/) is the standard.

**Reply by lishaduck** [reply]
Another tool. Still doesn't do what we want with navigation, but it tries: [modular](https://modular.flutterando.com.br/).

**Reply by lishaduck** [reply]
Some websites on structure:
1. [StackOverflow](https://stackoverflow.com/questions/56582329/how-to-organize-dart-files-inside-packages-in-flutter)
2. [Slidy](https://pub.dev/packages/slidy) (See <https://github.com/PSDTools/app/discussions/13#discussioncomment-5879507>)
3. [Dev.to](https://dev.to/davidadewoyin/structuring-flutter-project-26ob)
4. [Flutter Ready to Go (flavors, connectivity and more)](https://blog.juliobitencourt.com/2019/03/15/flutter-ready-to-go-flavors-connectivity-and-more/)
5. [GeeksForGeeks](https://www.geeksforgeeks.org/flutter-file-structure/)
6. [Bloc Pattern](https://medium.com/flutter-community/flutter-scalable-folder-files-structure-8f860faafebd)
7. [Flutter App Architecture with Riverpod: An Introduction](https://codewithandrea.com/articles/flutter-app-architecture-riverpod-introduction/)
8. [Testing W/out Mocks](https://www.jamesshore.com/v2/projects/nullables/testing-without-mocks)

_**Note**: I didn't read 5&6._

**Comment by lishaduck** [comment]
Off-topic, look at my edit on #19.

**Comment by lishaduck** [comment]
We're using Flutter, that's not debatable at this point. But, <https://framework7.io/> looks interesting if you (@ParkerH27 / @PetalCat) ever wanted to make something similar in the future.
